File: MyMovies-MVP/app/src/main/java/com/jbrunton/mymovies/search/SearchViewState.java
package com.jbrunton.mymovies.search;

import android.support.annotation.DrawableRes;

import com.google.auto.value.AutoValue;
import com.jbrunton.mymovies.Movie;

import java.util.List;

@AutoValue
public abstract class SearchViewState {
    public abstract List<SearchItemViewState> movies();
    public abstract boolean showError();
    public abstract String errorMessage();
    public abstract @DrawableRes int errorIcon();
    public abstract boolean showTryAgainButton();

    public static Builder builder() {
        return new AutoValue_SearchViewState.Builder()
                .setErrorIcon(0)
                .setErrorMessage("")
                .setShowTryAgainButton(false);
    }

    @AutoValue.Builder
    public abstract static class Builder {
        public abstract Builder setMovies(List<SearchItemViewState> movies);
        public abstract Builder setShowError(boolean showError);
        public abstract Builder setErrorMessage(String errorMessage);
        public abstract Builder setErrorIcon(@DrawableRes int errorIcon);
        public abstract Builder setShowTryAgainButton(boolean showTryAgainButton);
        public abstract SearchViewState build();
    }
}

File: MyMovies-MVP/app/src/main/java/com/jbrunton/mymovies/search/SearchViewState.kt
package com.jbrunton.mymovies.search

import android.support.annotation.DrawableRes
import com.jbrunton.mymovies.Movie
import com.jbrunton.mymovies.api.MaybeError
import java.util.*

data class SearchViewState(
        val movies: List<Movie>,
        val showError: Boolean,
        val errorMessage: String,
        val errorIcon: Int,
        val showTryAgainButton: Boolean = false)